doc/developer/design/20240925_network_policies.md

```diff
@@ -0,0 +1,148 @@
+# Network Policies
+
+- Associated:
+  - https://github.com/MaterializeInc/database-issues/issues/7062
+  - https://github.com/MaterializeInc/database-issues/issues/4637
+  - https://github.com/MaterializeInc/materialize/pull/29739
+  - https://github.com/MaterializeInc/materialize/pull/29179
+
+## The Problem
+Customers would like to restrict access to Materialize by IP address.
+https://github.com/MaterializeInc/database-issues/issues/4637
+
+## Success Criteria
+- Customers can define a global policy that restricts access to their Materialize environments based on the IP address of the client attempting to connect.
+- Materialize support can unlock an environment where policies prevent access.
+- The console is aware of whether an environment's network policies are blocking a connection it is trying to make.
+- Users can adjust network policies via SQL and in the console.
+
+Nice to haves:
+- Per-role network policies.
+- Policies for sources/sinks.
+- Mitigations to prevent user lockouts; such as ensuring the user is not blocking their current IP.
+- Termination of active connections based on newly applied policies.
+- Policies for egress traffic.
+
+## Out of Scope
+- Preventing out-of-policy traffic from reaching Environmentd.
+ (note this means network policies will not prevent DDOS)
+- Policy inheritance from associated roles. IE if 'bob' is a member of role 'eng'
+we will not apply policies from role 'eng' to 'bob'.
+- Restricting global API access.
+- Restricting access to Frontegg.
+
+## Solution Proposal
+
+#### Overview
+The proposed solution is to use role-based policies with a default network policy that applies to any role without a policy. This can initially be implemented as a global default policy and will be extended to per-user and per-source/sink. The policy will be applied when an attempt is made to establish a new client connection with the coordinator.
+
+#### New Resources
+A new `NetworkPolicy` resource will be added to the catalog.
+```rust
+struct NetworkPolicy {
+    id: NetworkPolicyId
+    name: String,
+    rules: Vec<NetworkPolicyRule>,
+}
+
+enum NetworkPolicyRule {
+    Ingress {
+        action: NetworkPolicyRuleAction,
+        source: IpNet,
+        comment: String
+ }
+}
```

Comment:
Will there ever be a case in Materialize where a single policy will be used in a context where there need to be both ingress and egress rules? With sources, we have push sources (webhooks) and pull sources (Kafka), but never both; it seems intuitive to me that a network policy on a webhook source would restrict inbound connections while a network policy on a Kafka source would restrict outbound connections.

If you buy this argument, I think concretely the difference would be using a generic term like `ADDRESSES` rather than `SOURCE` or `TARGET` or `DESTINATION` in network rules, and then whether the policy applied in the inbound or outbound direction would be a function of whether it was attached to a role, pull source, or push source.

```
    RULE (ACTION = ALLOW, ADDRESSES = 'cidr')
```

Reply:
I don't think you can intuit whether a policy/rule should be ingress or egress... two reasons:
1. If we want to introduce ingress sinks, then sinks would support both ingress and egress making it impossible to define a single policy default for sinks. 
2. We want to be able to share policies between resource types without having to ensure some color match. 

If we don't want ingress/egress enums, we could certainly handle this by having an ingress and egress rule set for policies. 


Thought for bullet 2
Let's say a user `bob` has a specific network policy `bob_policy`, that allows ingress access from IP `1.2.3.4/32`. 
It *should* follow that `bob` cannot create a source or sink that would expand their policy scope; i.e, any source or sink they create should inherit the policy applied to the user. This requires that the policy for `bob` be applicable to all resource types bob can create.

Reply:
yeah... I agree with you the network policies a user is able to associate with a given source/sink should be controlled by the policies on which the user has usage policies, not the users existing policies. If one wants to lock it down that only the policy of the user, they could grant the user only usage privileges to that policy.